src/bootstrap/bootstrap.py

```diff
@@ -437,10 +437,19 @@ def download_toolchain(self):
             filename = "cargo-{}-{}{}".format(rustc_channel, self.build,
                                             tarball_suffix)
             self._download_component_helper(filename, "cargo", tarball_suffix)
-            self.fix_bin_or_dylib("{}/bin/cargo".format(bin_root))
 
-            self.fix_bin_or_dylib("{}/bin/rustc".format(bin_root))
-            self.fix_bin_or_dylib("{}/bin/rustdoc".format(bin_root))
+            for bin_dir_name in ["bin", "libexec"]:
+                bin_dir = "{}/{}".format(bin_root, bin_dir_name)
+                for bin in os.listdir(bin_dir):
+                    bin_file = os.path.join(bin_dir, bin)
+
+                    # Skip script files, only actual executables need patching.
+                    with open(bin_file, "rb", buffering=0) as f:
+                        if f.read(2) == b"#!":
+                            continue
+
+                    self.fix_bin_or_dylib(bin_file)
```

Comment:
I mildly remember seeing this exact pattern before. Can’t tell if it was here, or in `nixpkgs`. Either way something like this seems fine to me, despite the apparent brittleness in the way patchable binaries are being detected. It would seem to me that an ideal approach would filter all ELF-looking files within `.` with the execute permissions and try to patch those, since that’s pretty much the intent behind this code (and would transparently cover both DSOs and executables alike…)

Reply:
To be clear, the shell filtering just removes `patchelf` printing a warning, it's otherwise a noop on non-ELF files, it's just an aesthetic fix for the output, I can remove the filtering entirely if that seems better.

(Not filtering would also mean easier to port the logic to Rust - which @jyn514 told me about, I had no idea this got duplicated)